### PR TITLE
service syncer

### PR DESCRIPTION
**Pull request by tomastigera**
## Description

Needed for programming routes for cluster IP towards the veth that would allow us to hook bpf into the path of host-networked processes and implement NAT for services.

    libcalico-go: Add service syncer
    
    Allow felix to monitor services and use them to program the dataplane.

    felix: Add Service update to calc

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

File: libcalico-go/lib/backend/k8s/resources/k8sservice.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+		kvps = append(kvps, kvp)
+		return &model.KVPairList{
+			KVPairs:  kvps,
+			Revision: revision,
+		}, nil
+	}
+
+	// Listing all services.
+	serviceList, err := c.clientSet.CoreV1().Services(rl.Namespace).List(ctx, metav1.ListOptions{ResourceVersion: revision})
```
Comment (caseydavenport) [inline libcalico-go/lib/backend/k8s/resources/k8sservice.go:120]: Take a look at our other backend resource clients - we should model this after them and use paged lists. 

Here's a parallel example for EndpointSlices: https://github.com/projectcalico/calico/blob/fc90dca3f652ade3fbadf3cc3ffd03bf20b136c4/libcalico-go/lib/backend/k8s/resources/kubeendpointslice.go#L93-L108

File: libcalico-go/lib/backend/model/resource.go [hunk 2]
```diff
@@ -148,6 +149,11 @@ func init() {
 		apiv3.KindKubeControllersConfiguration,
 		"kubecontrollersconfigurations",
 		reflect.TypeOf(apiv3.KubeControllersConfiguration{}))
+	registerResourceInfo(
+		KindK8sService,
+		"k8s-service",
```
Comment (caseydavenport) [inline libcalico-go/lib/backend/model/resource.go:154]: Should be "kubernetesservices" to match the other Kubernetes resources we pass through (annoying, but consistent)

File: libcalico-go/lib/backend/syncersv1/felixsyncer/felixsyncerv1.go [hunk 3]
```diff
@@ -93,6 +93,9 @@ func New(client api.Client, cfg apiconfig.CalicoAPIConfigSpec, callbacks api.Syn
 			additionalTypes = append(additionalTypes, watchersyncer.ResourceType{
 				ListInterface: model.ResourceListOptions{Kind: model.KindKubernetesEndpointSlice},
 			})
+			additionalTypes = append(additionalTypes, watchersyncer.ResourceType{
```
Comment (caseydavenport) [inline libcalico-go/lib/backend/syncersv1/felixsyncer/felixsyncerv1.go:96]: Commend on line 87 needs updating (was already out of date before this PR)

Comment (caseydavenport) [reply libcalico-go/lib/backend/syncersv1/felixsyncer/felixsyncerv1.go:96]: I assume we don't need to support whatever feature this is for in etcd mode?

File: libcalico-go/lib/namespace/resource.go [hunk 4]
```diff
@@ -25,6 +25,7 @@ const (
 	// to avoid an import loop.
 	KindKubernetesNetworkPolicy = "KubernetesNetworkPolicy"
 	KindKubernetesEndpointSlice = "KubernetesEndpointSlice"
+	KindK8sService              = "K8sService"
```
Comment (caseydavenport) [inline libcalico-go/lib/namespace/resource.go:28]: `KindKubernetesService` and `KubernetesService` to be consistent with other values here, please :grin:

**Review by tomastigera — COMMENTED** [review]
changed the naming

**Review by caseydavenport — APPROVED** [review]

**Comment by caseydavenport** [comment]
Looks like it has some conflicts @tomastigera

**Pull request merged**